tooling/cmd/promote-terraform/main.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	log.StandardLogger().SetLevel(logLevel(args))
+
+	signingEntity, err := loadSigningEntity(args.signingKeyText)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed decoding signing key")
+	}
+
+	files, err := downloadStagedArtifacts(context.Background(), args.providerTag, workspace.stagingDir, &args.staging)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed fetching artifacts")
+	}
+
+	objectStoreUrl := args.registryURL + "store/"
+
+	versionRecord, newFiles, err := repackProviders(files, workspace, objectStoreUrl, signingEntity, args.protocolVersions, args.providerNamespace, args.providerName)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed repacking artifacts")
+	}
+
+	err = updateRegistry(context.Background(), &args.production, workspace, args.providerNamespace, args.providerName, &versionRecord, newFiles)
+	if err != nil {
+		log.WithError(err).Fatal("Failed updating registry")
+	}
+}
+
+func logLevel(args *args) log.Level {
+	switch {
+	case args.verbosity >= 2:
+		return log.TraceLevel
+
+	case args.verbosity == 1:
+		return log.DebugLevel
+
+	default:
+		return log.InfoLevel
+	}
+}
+
+func updateRegistry(ctx context.Context, prodBucket *bucketConfig, workspace *workspacePaths, namespace, provider string, newVersion *registry.Version, files []string) error {
+	s3client, err := newS3ClientFromBucketConfig(ctx, prodBucket)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	versionsFileKey, versionsFilePath := makeVersionsFilePaths(workspace.registryDir, namespace, provider)
+	log.Infof("Downloading version index for %s/%s from %s", namespace, provider, versionsFileKey)
+
+	// Try downloading the versions file. This may not exist in an empty/new
+	// registry, so if the download fails with a NotFound error we ignore it
+	// and use an empty index.
+	versions := registry.NewVersions()
+	err = download(ctx, s3client, versionsFilePath, prodBucket.bucketName, versionsFileKey)
+	switch {
+	case err == nil:
+		versions, err = registry.LoadVersionsFile(versionsFilePath)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+	case isNotFound(err):
+		log.Info("No index found. Using empty index.")
+
+	default:
+		return trace.Wrap(err, "failed downloading index")
+	}
+
+	// Index the available version by their semver version, so that we can find the
+	// appropriate release if we're overwriting an existing version
+	versionIndex := map[semver.Version]*registry.Version{}
+	for _, v := range versions.Versions {
+		versionIndex[v.Version] = v
+	}
+
+	// add/overwrite the existing version entry
+	versionIndex[newVersion.Version] = newVersion
+	versions.Versions = flattenVersionIndex(versionIndex)
+
+	if err = versions.Save(versionsFilePath); err != nil {
+		return trace.Wrap(err, "failed saving index file")
+	}
+	files = append(files, versionsFilePath)
+
+	// Finally, push the new index files to the production bucket
+	err = uploadRegistry(ctx, s3client, prodBucket.bucketName, workspace.productionDir, files)
```

Comment:
Is there anything we can do on the tool side to make sure that the index file hasn't changed between when we downloaded it above and saving it here? In case we get simultaneous builds running somehow. I don't suppose S3 has something like compare-and-swap?